### PR TITLE
[feat] Weights and Biases Logging Feature

### PR DESCRIPTION
**Pull request by mudit-dhawan**
Add a WandbLogger class to support wandb logging. 
Update config options to enable wandb logger and integrate with the trainer.
To enable wandb logging use the following options `training.wandb=True  
env.wandb_projectname=<ProjectName> env.wandb_runname=<RunName>`

**Examples:**
Screenshot of the wandb dashboard for two runs of m4c project on textvqa dataset.  
image
1. command line code for _run1_: `mmf_run config=projects/m4c/configs/textvqa/defaults.yaml datasets=textvqa model=m4c run_type=train_val training.batch_size=2 training.log_interval=2 training.wandb=True training.max_updates=10 env.wandb_projectname=Try1 env.wandb_runname=run1 `
2. command line code for _run2_: `mmf_run config=projects/m4c/configs/textvqa/defaults.yaml datasets=textvqa model=m4c run_type=train_val training.batch_size=2 training.log_interval=1 training.wandb=True training.max_updates=10 env.wandb_projectname=Try1 env.wandb_runname=run2 `

 Linked Issue: Wandb Logger in mmf #588

### CODE REVIEW

File: mmf/utils/wandb_logger.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+        self._is_master = is_master()
+
+        self.setup()
+
+    def setup(self):
+        """
+        Setup `Weights and Biases` for logging.
+        """
+        if self._is_master:
```
Comment (vedanuj) [inline mmf/utils/wandb_logger.py:56]: if is_master()
```suggestion
        if is_master():
```

File: mmf/utils/wandb_logger.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+                self._wandb.define_metric(
+                    "*", step_metric="trainer/global_step", step_sync=True
+                )
+
+    def __del__(self):
+        if getattr(self, "_wandb", None) is not None:
+            self._wandb.finish()
+
+    def _should_log_wandb(self):
+        if self._wandb is None or not self._is_master:
```
Comment (vedanuj) [inline mmf/utils/wandb_logger.py:73]: Same as above

**Review by vedanuj — CHANGES_REQUESTED** [review]
Thanks for contributing to MMF. Requesting some changes before this is ready to land. 

Also will you be interested to add a doc under the `notes` section [here](https://github.com/facebookresearch/mmf/tree/master/website/docs/notes)  called `logging.md` and add a section detailing how to use WandDB logger? Will be super useful for users if this information for usage is documented.

File: mmf/utils/wandb_logger.py (hunk 1 above)
Comment (mudit-dhawan) [reply mmf/utils/wandb_logger.py]: Fixed. Do let me know if any further changes are required.

File: mmf/configs/defaults.yaml [hunk 3]
```diff
@@ -285,6 +288,21 @@ env:
...
+    # Log directory for Weights and Biases, default points to same as logs
+    # Only used when training.wandb is enabled.
+    # Use MMF_WANDB_LOGDIR or env.wandb_logdir to override
+    wandb_runname: ${env:MMF_WANDB_RUNNAME,}
+
+    # Log directory for Weights and Biases, default points to same as logs
+    # Only used when training.wandb is enabled.
+    # Use MMF_WANDB_LOGDIR or env.wandb_logdir to override
+    wandb_logdir: ${env:MMF_WANDB_LOGDIR,}
+
```
[the remark this answers is not in the document]
Comment (mudit-dhawan) [reply mmf/configs/defaults.yaml]: Fixed. 

Now to enable wandb logging one would use the following options: `training.wandb.enabled=True training.wandb.wandb_projectname=<ProjectName> training.wandb.wandb_runname=<RunName>`

File: mmf/utils/wandb_logger.py (hunk 2 above)
Comment (mudit-dhawan) [reply mmf/utils/wandb_logger.py]: Fixed.

**Comment by mudit-dhawan** [comment]
Hey @vedanuj , have added `logging.md`  under the `notes` section [here](https://github.com/facebookresearch/mmf/tree/master/website/docs/notes), with a section `Weights and Biases Logger` detailing how to use Wandb Logging feature.

File: mmf/utils/wandb_logger.py [hunk 4]
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
```
Comment (apsdehal) [inline mmf/utils/wandb_logger.py:1]: We should move this to `mmf/utils/logger.py` to keep things around logger together.

File: mmf/utils/wandb_logger.py [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+from typing import Dict, Optional
+
+from mmf.utils.distributed import is_master
+
+
+try:
+    import wandb
```
Comment (apsdehal) [inline mmf/utils/wandb_logger.py:9]: Move this check inside the init of `WandbLogger`. In exception, you can then throw a warning to install wandb and die.

File: mmf/configs/defaults.yaml [hunk 6]
```diff
@@ -41,6 +41,16 @@ training:
...
 
+    # Weights and Biases control, by default Weights and Biases (wandb) is disabled
+    wandb: 
+        # Whether to use Weights and Biases Logger, (Default: false)
+        enabled: false
+        # Project name to be used while logging the experiment with wandb
+        wandb_projectname: project
+        # Experiment/ run name to be used while logging the experiment 
+        # under the project with wandb
+        wandb_runname: run
```
Comment (apsdehal) [inline mmf/configs/defaults.yaml:52]:
```suggestion
        wandb_runname: ${training.experiment_name}
```

File: mmf/configs/defaults.yaml [hunk 7]
```diff
@@ -41,6 +41,16 @@ training:
     # Tensorboard control, by default tensorboard is disabled
     tensorboard: false
 
+    # Weights and Biases control, by default Weights and Biases (wandb) is disabled
+    wandb: 
+        # Whether to use Weights and Biases Logger, (Default: false)
+        enabled: false
+        # Project name to be used while logging the experiment with wandb
+        wandb_projectname: project
```
Comment (apsdehal) [inline mmf/configs/defaults.yaml:49]:
```suggestion
        wandb_projectname: mmf_${oc.env:USER}
```

This might be better to be more specific.

**Review by apsdehal — CHANGES_REQUESTED** [review]
Thanks for making this awesome change. Minor changes and then we should be ready to land this.

File: mmf/utils/wandb_logger.py (hunk 4 above)
Comment (mudit-dhawan) [reply mmf/utils/wandb_logger.py]: Fixed.

File: mmf/utils/wandb_logger.py (hunk 5 above)
Comment (mudit-dhawan) [reply mmf/utils/wandb_logger.py]: Fixed.

**Comment by apsdehal** [comment]
@mudit-dhawan Can you resolve the conflicts so we can land this?

**Comment by mudit-dhawan** [comment]
> @mudit-dhawan Can you resolve the conflicts so we can land this?

Thanks for the reminder @apsdehal , sorry for the delay. I have resolved the merging conflict. Please let me know if any further changes are required.  Thanks.

**Review by apsdehal — APPROVED** [review]

File: mmf/utils/logger.py [hunk 8]
```diff
@@ -383,3 +388,78 @@ def add_histogram_for_model(self, model, iteration):
...
+
+        self._wandb_init.update(**init_kwargs)
+
+        self.setup()
+
+    def setup(self):
+        """
+        Setup `Weights and Biases` for logging.
+        """
+        if _is_master():
```
Comment (vedanuj) [inline mmf/utils/logger.py:433]: I think this should be `is_master()`. flake8 test are failing due to this.

File: mmf/utils/logger.py [hunk 9]
```diff
@@ -383,3 +388,78 @@ def add_histogram_for_model(self, model, iteration):
...
+                self._wandb.define_metric(
+                    "*", step_metric="trainer/global_step", step_sync=True
+                )
+
+    def __del__(self):
+        if getattr(self, "_wandb", None) is not None:
+            self._wandb.finish()
+
+    def _should_log_wandb(self):
+        if self._wandb is None or not _is_master():
```
Comment (vedanuj) [inline mmf/utils/logger.py:450]: Same for this.

**Review by apsdehal — APPROVED** [review]

**Comment by g-luo** [comment]
I would recommend adding the following two steps to the docs / linking to wandb setup (https://docs.wandb.ai/quickstart):

```
pip install wandb
wandb login
```

I was running and it wasn't logging because I hadn't set up wandb before, and for some reason it also wasn't outputting any error messages.

**Pull request closed without merging**